apis/placement/v1alpha1/stagedupdate_types.go

```diff
@@ -0,0 +1,441 @@
+/*
+Copyright (c) Microsoft Corporation.
+Licensed under the MIT license.
+*/
+
+package v1alpha1
+
+import (
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+
+	"go.goms.io/fleet/apis/placement/v1beta1"
+)
+
+// +genclient
+// +genclient:namespaced
+// +kubebuilder:object:root=true
+// +kubebuilder:subresource:status
+// +kubebuilder:resource:scope="Namespaced",categories={fleet,fleet-placement}
+// +k8s:deepcopy-gen:interfaces=k8s.io/apimachinery/pkg/runtime.Object
+
+// StagedUpdateRun defines a stage by stage update run that applies the selected resources by the
+// corresponding ClusterResourcePlacement to its selected clusters. We remove the resources from the clusters that are
+// unselected after all the stages explicitly defined in the updateStrategy complete.
+// Each StagedUpdateRun object corresponds to a single "release" of a certain version of the resources.
+// The release is abandoned if the StagedUpdateRun object is deleted or the scheduling decision changes.
+type StagedUpdateRun struct {
+	metav1.TypeMeta   `json:",inline"`
+	metav1.ObjectMeta `json:"metadata,omitempty"`
+
+	// The desired state of StagedUpdateRun. The spec is immutable.
+	// +kubebuilder:validation:Required
+	// +kubebuilder:validation:XValidation:rule="self == oldSelf",message="The spec field is immutable"
+	Spec StagedUpdateRunSpec `json:"spec"`
+
+	// The observed status of StagedUpdateRun.
+	// +kubebuilder:validation:Optional
+	Status StagedUpdateRunStatus `json:"status,omitempty"`
+}
+
+// StagedUpdateRunSpec defines the desired the update strategy and the snapshot indices
+// on which the update is based.
```

Comment:
// StagedUpdateRunSpec contains the snapshot index of the selected resources to be updated, and the stage-by-stage strategy to update them across selected cluster, for the specified ResourcePlacement object.